ext/opcache/zend_persist.c

```diff
@@ -190,6 +190,9 @@ static zend_ast *zend_persist_ast(zend_ast *ast)
 		node = (zend_ast *) copy;
 	} else if (ast->kind == ZEND_AST_OP_ARRAY) {
 		zend_ast_op_array *copy = zend_shared_memdup(ast, sizeof(zend_ast_op_array));
+		/* We're holding a separate reference to the op_array in the AST. Release it
+		 * early because zend_persist_op_array() is destructive. */
+		destroy_op_array(copy->op_array);
 		zval z;
 		ZVAL_PTR(&z, copy->op_array);
```

Comment:
This looks strange. You first destroy and then use. May be you should use the ``copy``.

Reply:
The comment above should explain it. Persist assumes rc=1 and partially destroys the op_array. After that point, we can't call destroy_op_array() anymore. We do it before, but we know a reference will always be remaining.

Reply:
Sure, that would make sense.

Reply:
Looking at `destroy_op_array` and `zend_persist_op_array_ex()`, perhaps an alternative and possibly cleaner fix would be not handling the refcount of the `function_name` separately from the refcount of the `op_array`: No matter how many references there are to the `op_array`, there is only one `op_array` holding onto the `function_name` and thus there should only be 1 reference added for the use in that `op_array`, no?

Reply:
In this case, yes. But it's also used for traits, where we have different `op_array`s pointing to the same `opcodes` arrays. You can probably use your own function that only increments refcount, but then you also can't use `destroy_op_array()` because it decrements the function name RC.

Reply:
> In case AST would keep index of op_array in `op_array.dynamic_func_defs[]` everything could work without this magic out of the box.

Ilija suggested in private that we might not need the entry in `dynamic_func_defs` at all and it appears he is correct. I've updated the PR and locally all the tests pass without OPcache, with OPcache and also a standalone test with file cache works.